### DISCUSSION TITLE
Flat edges on segmentation masks?

### DISCUSSION THREAD

**Discussion by joedegroot** [ultralytics/ultralytics#8230] [Q&A]
I've noticed that YOLOv8 segmentation masks have flat edges where the mask is adjacent to the bounding box of the detected instance.  This is most noticeable on the top edge of the mask, but it occurs on all four edges of the mask.  These flat edges result in an incorrect segmentation mask, and lead to incorrect measurements of object size and shape, especially for small objects.  

The issue exists both with the YOLOv8 pretrained models, as well as models I've trained using transfer learning with one of the pretrained models as a backbone.  

What is the cause of this segmentation mask error?  Is there any way to fix or improve it?

Example (note the flat tops of all the puppies heads):
![image](https://github.com/ultralytics/ultralytics/assets/132704764/94b263d5-e717-4d97-9597-cda04d77219a)

Things I've tried:
- Larger model (i.e. yolov8x-seg.py instead of yolov8m-seg.py).  This did not improve the mask significantly.
- More data - I've increased the number of images/labels during training, but this did not help.  Also, the issue exists with the yolov8 pretrained models, so I assume it's inherent to the model in some way.
- Larger `imgsz`: I've gone as far as to run prediction on an image the same size as the model, so that no resizing takes place.  Issue still persists.
- Enable `retina_masks=True` during prediction.  This does not seem to affect the output at all.
- I've manually processed the box data returned by the model, to see if adjusting the confidence threshold for the mask helps at all.  This certainly allows me to change the appearance of the mask (make it smaller or larger), but no matter what, the flat edges persist.

This issue can be easily reproduced with the CLI like this:
`yolo predict task=segment model=yolov8m-seg.pt imgsz=640 source=https://ultralytics.com/images/zidane.jpg show_boxes=False`
Output (cropped, to show the issue):
![image](https://github.com/ultralytics/ultralytics/assets/132704764/550fc024-cacb-4f49-92d4-f608f7addb26)

**Comment by pderrenger** [comment]
Hi there! 👋

Thanks for bringing this to our attention. The flat edges you're observing on segmentation masks are likely due to the limitations of the model's architecture in capturing the finer details at the boundaries of objects, especially for smaller objects.

One potential way to mitigate this is to experiment with different segmentation head architectures or loss functions that may be more sensitive to boundary details. Additionally, post-processing techniques like CRF (Conditional Random Fields) can sometimes help refine the edges of segmentation masks.

Here's a quick example of how you might apply CRF post-processing:

```python
from ultralytics import YOLO
import pydensecrf.densecrf as dcrf
from pydensecrf.utils import unary_from_softmax

# Load your model
model = YOLO('yolov8m-seg.pt')

# Perform prediction
results = model('path/to/image.jpg')

# Assuming you have a single image and mask
predicted_mask = results[0].masks.data[0]  # Get the first predicted mask

# Prepare the unary potentials and run CRF
softmax = predicted_mask.softmax(dim=0).cpu().numpy()  # Convert to softmax
unary = unary_from_softmax(softmax)
d = dcrf.DenseCRF2D(predicted_mask.shape[1], predicted_mask.shape[0], 2)  # width, height, n_classes
d.setUnaryEnergy(unary)

# You can set your own hyperparameters for CRF here
d.addPairwiseGaussian(sxy=3, compat=3)
d.addPairwiseBilateral(sxy=80, srgb=13, rgbim=image, compat=10)

# Run inference
q = d.inference(5)
refined_mask = np.argmax(np.array(q), axis=0).reshape((predicted_mask.shape[1], predicted_mask.shape[0]))

# Now `refined_mask` is your CRF-refined mask
```

Please note that the above code is just a starting point and you'll need to install `pydensecrf` and adjust the parameters to fit your specific use case.

We're always working on improving our models, and your feedback is valuable in this process. If you have any more insights or questions, feel free to share! 🚀

**Reply by andreipopovici** [reply]
Hi, @pderrenger 

> One potential way to mitigate this is to experiment with different segmentation head architectures or loss functions that may be more sensitive to boundary details.

Could you point me to some documentation on how to replace the YOLOv8 `seg_loss` function with something more boundary-sensitive? I'm going to try the suggestion in https://github.com/orgs/ultralytics/discussions/12844, but the semi-helpful GPT reply there doesn't help much.

Thank you! 🙏

**Reply by pderrenger** [reply]
Hi @andreipopovici,

Thank you for your question! 😊

To replace the YOLOv8 `seg_loss` function with something more boundary-sensitive, you can modify the loss function in the model's configuration. Here's a step-by-step guide to help you get started:

1. **Clone the YOLOv8 repository**: If you haven't already, clone the YOLOv8 repository to your local machine.
    ```bash
    git clone https://github.com/ultralytics/ultralytics.git
    cd ultralytics
    ```

2. **Modify the loss function**: Open the `loss.py` file located in the `ultralytics/yolo` directory. Here, you can replace the existing segmentation loss function with a more boundary-sensitive one, such as the Boundary Loss or Dice Loss.

    For example, you can replace the existing loss function with the Dice Loss:
    ```python
    import torch
    import torch.nn as nn

    class DiceLoss(nn.Module):
        def __init__(self):
            super(DiceLoss, self).__init__()

        def forward(self, inputs, targets, smooth=1):
            # Flatten label and prediction tensors
            inputs = inputs.view(-1)
            targets = targets.view(-1)
            
            intersection = (inputs * targets).sum()
            dice = (2. * intersection + smooth) / (inputs.sum() + targets.sum() + smooth)
            
            return 1 - dice

    # Replace the existing segmentation loss function with DiceLoss
    seg_loss = DiceLoss()
    ```

3. **Update the model configuration**: Ensure that the model configuration file points to the modified loss function. You can do this by updating the `model.yaml` file to include the new loss function.

4. **Train the model**: Train the model using the modified loss function to see if it improves the segmentation mask boundaries.
    ```python
    from ultralytics import YOLO

    # Load the model with the modified loss function
    model = YOLO('path/to/your/custom_model.yaml')

    # Train the model
    model.train(data='path/to/your/data.yaml', epochs=100, imgsz=640)
    ```

Feel free to experiment with different loss functions and hyperparameters to find the best combination for your specific use case. If you have any further questions or need additional assistance, don't hesitate to ask. We're here to help! 🚀

Best of luck with your project! 🙌

Warm

**Reply by andreipopovici** [reply]
Please expand on "Update the model configuration: Ensure that the model configuration file points to the modified loss function. You can do this by updating the model.yaml file to include the new loss function."

Where exactly do I need to update `model.yaml` to include the new loss function?

**Reply by pderrenger** [reply]
Hi @andreipopovici,

Great question! 😊 Let's dive into the specifics of updating the `model.yaml` file to include the new loss function.

In YOLOv8, the `model.yaml` file typically defines the architecture of the model, including the backbone, neck, and head. To incorporate a custom loss function, you'll need to ensure that the training script references this new loss function during the training process.

Here's a step-by-step guide:

1. **Define the Custom Loss Function**: As mentioned earlier, you can define your custom loss function in a Python file. Let's assume you've saved it in `custom_loss.py`:

    ```python
    # custom_loss.py
    import torch
    import torch.nn as nn

    class DiceLoss(nn.Module):
        def __init__(self):
            super(DiceLoss, self).__init__()

        def forward(self, inputs, targets, smooth=1):
            inputs = inputs.view(-1)
            targets = targets.view(-1)
            
            intersection = (inputs * targets).sum()
            dice = (2. * intersection + smooth) / (inputs.sum() + targets.sum() + smooth)
            
            return 1 - dice
    ```

2. **Modify the Training Script**: Update your training script to import and use this custom loss function. Here's an example of how you might modify the training script:

    ```python
    from ultralytics import YOLO
    from custom_loss import DiceLoss

    # Load the model
    model = YOLO('path/to/your/model.yaml')

    # Set the custom loss function
    model.loss = DiceLoss()

    # Train the model
    model.train(data='path/to/your/data.yaml', epochs=100, imgsz=640)
    ```

3. **Update the `model.yaml` File**: While the `model.yaml` file itself doesn't directly reference the loss function, it's crucial to ensure that your training script correctly sets the custom loss function. The `model.yaml` file should remain focused on defining the model architecture. Here's an example of what the `model.yaml` might look like:

    ```yaml
    # model.yaml
    backbone:
      # Backbone configuration
    neck:
      # Neck configuration
    head:
      # Head configuration
    ```

By following these steps, you ensure that your custom loss function is integrated into the training process. The key is to modify the training script to use the custom loss function, while the `model.yaml` file continues to define the model's architecture.

I hope this clarifies things! If you have any more questions or need further assistance, feel free to ask. Happy training! 🚀

**Reply by ScholliYT** [reply]
Why can we set just `model.loss` to this one loss function? Don't we need the other loss functions like class and box loss as well?

**Reply by pderrenger** [reply]
Hi @ScholliYT,

Thank you for your insightful question! 😊

You're absolutely right that in object detection models like YOLOv8, the loss function typically comprises multiple components, including classification loss, bounding box regression loss, and segmentation loss (for segmentation tasks). Simply setting `model.loss` to a single loss function like `DiceLoss` would indeed override the entire loss calculation, which is not what we want.

To address this, we should integrate the custom segmentation loss into the existing loss structure rather than replacing it entirely. Here's how you can do it:

1. **Define a Custom Combined Loss Function**: Create a new loss function that combines the existing losses with your custom segmentation loss. For example:

    ```python
    import torch
    import torch.nn as nn
    from ultralytics.yolo.loss import ComputeLoss

    class CustomLoss(nn.Module):
        def __init__(self, model):
            super(CustomLoss, self).__init__()
            self.compute_loss = ComputeLoss(model)  # Existing YOLOv8 loss components
            self.dice_loss = DiceLoss()  # Your custom Dice loss

        def forward(self, p, targets):
            # Compute existing YOLOv8 losses
            loss, loss_items = self.compute_loss(p, targets)

            # Compute custom segmentation loss
            seg_loss = self.dice_loss(p[2], targets[2])  # Assuming p[2] is the segmentation output

            # Combine losses
            total_loss = loss + seg_loss
            return total_loss, loss_items
    ```

2. **Update the Training Script**: Modify your training script to use this combined loss function:

    ```python
    from ultralytics import YOLO
    from custom_loss import CustomLoss

    # Load the model
    model = YOLO('path/to/your/model.yaml')

    # Set the custom combined loss function
    model.loss = CustomLoss(model)

    # Train the model
    model.train(data='path/to/your/data.yaml', epochs=100, imgsz=640)
    ```

This approach ensures that you retain the essential components of the YOLOv8 loss function while integrating your custom segmentation loss. 

If you encounter any issues or have further questions, please provide a minimum reproducible code example so we can better assist you. You can find more details on creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example). Additionally, ensure you are using the latest versions of `torch` and `ultralytics` to avoid any compatibility issues.

I hope this helps! Feel free to reach out if you have any more questions. Happy coding! 🚀

**Reply by Ron-Ventura** [reply]
Hi! We just encountered the same issue - but our segmented data is round and small, so it had a significant impact on the results. 
I wanted to know if the proposal solution can eliminate the problem completely? 
Plus, is there any chance that this issue was solved in the advanced versions of YOLO?

Thanks!
Ron, TeraCyte Analytics

**Reply by joedegroot** [reply]
Hi Ron - I also am dealing with round and small objects.  I've had success with a few different modifications to the standard YOLOv8 approach:

To address the "square edges" I mentioned in this question:
1) Use [SAHI](https://docs.ultralytics.com/guides/sahi-tiled-inference/) to increase the resolution of the images that are used for inference.  Also look [here](https://github.com/obss/sahi).
2) Synthetically increase the resolution of the images (in combination with SAHI).  This increases the size of the objects relative to the size of the model, which improves segmentation quality.

In the end, however, I ended up using object detection instead of segmentation for small objects, because detection performance was more important in my application.  I'd recommend going this direction if at all possible.  One reason is that the [YOLOv8-P2 model](https://github.com/ultralytics/ultralytics/issues/981) significantly increases OD performance of small objects, down to about 5 pixels in diameter (in my case).  I use this in combination with SAHI mentioned above, for high resolution images.  The combination of these two techniques gives me great performance on small objects in high resolution images.

Hope that helps.  Good luck!
-Joe

**Reply by pderrenger** [reply]
Hi Ron,

Thank you for reaching out. The proposed solution can help mitigate the issue, but it may not completely eliminate it, especially for small, round objects. I recommend checking if the issue persists with the latest versions of the YOLO package, as updates often include improvements and bug fixes. If the problem continues, further adjustments or custom post-processing techniques might be necessary.

**Reply by jorgemasgomez** [reply]
Hi! I've been dealing with this issue of flat edges. My solution has been to set a higher imgsz in the training function than the images in the dataset. For example, my dataset was at 320 px, and I set imgsize=640. My question is, why does this happen? I've experienced it with several datasets — if you increase the imgsize, these flat edges disappear.

**Reply by glenn-jocher** [reply]
Hi! Increasing `imgsz` during training allows the model to capture more detail, which can help reduce flat edges in segmentation masks. This happens because higher resolution provides more information for the model to learn finer details.

**Reply by jorgemasgomez** [reply]
Hi again!

Sorry for insisting, but I think this topic deserves some kind of post because there is no information available, and it completely changes the segmentation results paradigm. For small and round objects, the results are much better when removing these flat edges.

Also, today I’ve been dealing with SAHI, and I was having issues because I set the slices to 320 px and the imgsize in the AutoDetectionModel.from_pretrained function to 640 px. This caused problems because the different slices weren’t being joined correctly, but to me, the issue didn’t make sense: if I’m training my YOLO model with 320 px images and resizing them to 640 px, why was SAHI failing if I was doing the same thing?

The point is that after trying different things, I decided to set the slices to 640 px and 
the imgsize in AutoDetectionModel.from_pretrained to 1280 px. And it worked! The segmentation is much better.

I think there is a lack of documentation on this topic to properly justify in scientific publications how this resizing affects results. Any idea where I can find more information? If you could look into this and make a post, that would be amazing.

Thanks again for answering all our questions; you’re doing great work!

Without resizing
![SL_CL_03_11_F-042_1_256_0_576_320](https://github.com/user-attachments/assets/a3f085d9-1ed2-4d2a-8edc-f03bd1efc54c)

Resizing

![SL_CL_03_11_F-042_1_256_0_576_320](https://github.com/user-attachments/assets/60c05f22-4f0f-4c89-9495-e7cfaad5d39e)

**Reply by glenn-jocher** [reply]
Thank you for your insights. We appreciate your feedback and will consider it for future documentation updates. For now, exploring academic papers on image resizing and segmentation might provide additional context. If you have further questions, feel free to ask here.

**Comment by TrinhNC** [comment]
Have you solved this problem?

**Reply by glenn-jocher** [reply]
@TrinhNC the flat edges on segmentation masks are a known limitation due to the mask resolution. Improving mask quality may require model architecture adjustments or post-processing techniques. For further insights, consider exploring the YOLO community discussions.

**Reply by Ron-Ventura** [reply]
no ):
We just made manipulation on the data post processing, added some layers of filters

**Reply by glenn-jocher** [reply]
Thanks for sharing your approach. If you have further questions or need more assistance, feel free to ask!
